File: s-api-opm.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//DEPS jakarta.json.bind:jakarta.json.bind-api:3.0.1
//DEPS org.eclipse:yasson:3.0.3
//DEPS jakarta.json:jakarta.json-api:2.1.3
//DEPS commons-io:commons-io:2.16.1

// mvn dependenncy:get -DgroupId=jakarta.json.bind -DartifactId=jakarta.json.bind-api -Dversion=3.0.1
// mvn dependency:get -DgroupId=org.eclipse -DartifactId=yasson -Dversion=3.0.3
// mvn dependency:get -DgroupId=jakarta.json -DartifactId=jakarta.json-api -Dversion=2.1.3

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.SocketException;
import java.net.URL;

import java.util.List;
import java.util.Map;
import java.util.Base64;

import java.time.LocalDateTime;

import java.io.FileWriter;
import java.io.IOException;
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.Path;

import java.util.regex.Matcher;
import java.util.regex.Pattern;


import java.io.FileOutputStream;
import java.io.InputStream;

import jakarta.json.bind.Jsonb;
import jakarta.json.bind.JsonbBuilder;
import jakarta.json.bind.JsonbBuilder;

import org.apache.commons.io.FileUtils;

public class SApiOpm {

    private static String opmURL = "https://gist.githubusercontent.com/funkyhippo/1d40bd5dae11e03a6af20e5a9a030d81/raw/?";

    public static void main(String... args) {
        try {
            var response = getRequest();
            if (response == null) {
                System.out.println("Something went wrong with response, exiting!");
                System.exit(1);
            }

            Jsonb jsonb = JsonbBuilder.create();
            Opm opm = jsonb.fromJson(response, Opm.class);

            printInfo(opm);
           getPics(opm.chapters);
        }
        catch(jakarta.json.bind.JsonbException je) {
            System.out.println(je);
            je.printStackTrace();
        }
        catch (Exception e) {
            System.out.println(e);
            e.printStackTrace();
        }

    }

    private static void getPics(Map<String, Chapter> chapters) {
        var home = System.getProperty("user.home");
        chapters.forEach((k, v) -> {
                var chapterDir = Paths
                    .get(home, "Downloads", "opm", k)
                    .toString()
                    .replace(".", "_");

                System.out.println();
                System.out.println("Chapter: " + k);
                System.out.printf("%-2sDirectory: %s\n", "", chapterDir);

                v.groups.get("/r/OnePunchMan").forEach(link -> {
                        try {
                            Thread.sleep(3000);
                            var finalLink = getDecodedLink(link);
                            var picName = new String(URI.create(finalLink).getPath()).substring(1);
                            var filepath = Paths.get(chapterDir, picName);

                            if(!Files.exists(filepath)) {
                                System.out.printf("%-4surl: %s - filepath: %s\n", "", finalLink, filepath);
                                try {
                                    Files.createDirectories(Paths.get(chapterDir));

                                    var finalLinkUrl = new URL(finalLink).openConnection();
                                    finalLinkUrl.setRequestProperty("User-Agent", "Mozilla/5.0");
                                    FileUtils.copyInputStreamToFile(finalLinkUrl.getInputStream(),
                                                                    new File(filepath.toString()));
                                } catch(SocketException se) {
                                    System.out.println("Error: " + se.getMessage());
                                    se.printStackTrace();
                                }
                                catch(IOException ioe) {
                                    System.out.println("Error: error copying file: " + ioe.getMessage());
                                    ioe.printStackTrace();
                                }
                            }
                        } catch(InterruptedException e){
                            System.out.println(e.getMessage());
                            e.printStackTrace();
                        }
                    });
            });
    }

    private static String chapterZeroAppend(String meh) {
        String regex = "\\b(\\d)_";
        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(meh);
        if(matcher.find()) {
            return "0" + meh;
        }

        return meh;
    }

    private static void printInfo(Opm opm) {
        System.out.printf("\ntitle: %s\n", opm.title);
        System.out.printf("latest cover: %s\n", getDecodedLink(opm.cover));
    }

	private static String getRequest() {
      var jsonLocation = "/tmp/opm"+ ".json"; // + LocalDateTime.now()
      var file = new File(jsonLocation);
      if(file.exists()) {
          System.out.println("WARNING: JSON file already downloaded, imma using it!");
          try {
              return String.join(System.lineSeparator(), Files.readAllLines(Paths.get(jsonLocation)));
          } catch(IOException ioe) {
              System.out.println("An error occurred while saving the JSON data: " + ioe.getMessage());
              ioe.printStackTrace();
}
      }

      HttpResponse<String> response = null;
		var client = HttpClient.newHttpClient();
      var request = HttpRequest
            .newBuilder()
            .uri(URI.create(opmURL))
            .build();

        try {
            response = client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch(InterruptedException ie) {
            System.out.println("An error occurred while saving the JSON data: " + ie.getMessage());
            ie.printStackTrace();
        } catch (IOException e) {
            System.out.println("An error occurred while saving the JSON data: " + e.getMessage());
            e.printStackTrace();
        }

        try {
            var filew = new FileWriter(jsonLocation);
            filew.write(response.body());
            filew.close();
        } catch(IOException ioe) {
            System.out.println("An error occurred while saving the JSON data: " + ioe.getMessage());
            ioe.printStackTrace();
        }

        return response.body();
	}

    private static String getDecodedLink(String link) {
        String result = null;
        var uLink = URI.create(link).getPath().replaceAll("/v1/image/", "");
        result = new String(Base64.getDecoder().decode(uLink));

        return result;
    }

    public record Chapter(
                          Map<String, List<String>> groups,
                          long lastUpdated,
                          String title,
                          String volume) {
    }

    public record Opm(
                      String cover,
                      String description,
                      String title,
                      Map<String, Chapter> chapters) {
    }

}

File: s_api_opm.go
/// 2>/dev/null ; gorun "$0" "$@" ; exit $?

package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"strings"
	"time"
)

const opmURL = "https://gist.githubusercontent.com/funkyhippo/1d40bd5dae11e03a6af20e5a9a030d81/raw/?"

func main() {
	response := parseResponse()
	printInitialInfo(response)
	getPics(response.Chapters)
}

func printInitialInfo(response Opm) {
	fmt.Println("cover:", getDecodedLink(response.Cover))
	fmt.Println("title:", response.Title)
}

func destDir() string {
	if len(os.Args) == 2 {
		return os.Args[1]
	}

	home := home()
	result := filepath.Join(home, "Downloads", "opm")

	return result
}

func getPics(chapters map[string]Chapter) {
	for k, v := range chapters {
		chapterDir := strings.Replace(
			filepath.Join(destDir(), k),
			".", "_", -1)

		fmt.Println("\nChapter:", k)
		fmt.Printf("%-2sDirectory: %s\n", "", chapterDir)

		for _, link := range v.Groups["/r/OnePunchMan"] {
			finalUrl := getDecodedLink(link)
			picNameParsed, err := url.Parse(finalUrl)
			if err != nil {
				fmt.Println("Error parsing image url", err)
				os.Exit(1)
			}
			picName := picNameParsed.Path[1:]
			destinationPath := filepath.Join(chapterDir, picName)

			if _, err := os.Stat(destinationPath); err == nil {
				fmt.Printf("%-4sfound: %s\n", "", destinationPath)
				continue
			}

			time.Sleep(5 * time.Second)

			err = os.MkdirAll(filepath.Dir(destinationPath), 0755)
			if err != nil {
				fmt.Println("Error: unable to create parents dirs of:", destinationPath, err)
				os.Exit(1)
			}

			fmt.Printf("%-4surl: %s - filepath: %s\n", "", finalUrl, destinationPath)
			err = downloadImage(finalUrl, destinationPath)
			if err != nil {
				fmt.Println("Error: unable to download file:", destinationPath, err)
			}
		}
	}
}

func downloadImage(url, destinationPath string) error {
	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyFromEnvironment,
		},
	}

	// Create a new HTTP request with the specified URL
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return err
	}

	// Set the User-Agent header
	userAgent := "Mozilla/5.0"
	req.Header.Set("User-Agent", userAgent)

	// Send the HTTP request
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Create the destination file
	file, err := os.Create(destinationPath)
	if err != nil {
		return err
	}
	defer file.Close()

	// Write the image data to the file
	_, err = io.Copy(file, resp.Body)
	if err != nil {
		return err
	}

	return nil
}

func home() string {
	result, err := os.UserHomeDir()
	if err != nil {
		fmt.Println("Error getting user home", err)
		os.Exit(1)
	}

	return result
}

func getDecodedLink(link string) string {
	var result string
	uLink, err := url.Parse(link)
	if err != nil {
		fmt.Println("Error parsing URI", link, err)
		os.Exit(1)
	}

	result = path.Base(uLink.Path)
	decoded, err := base64.StdEncoding.DecodeString(result)
	if err != nil {
		fmt.Println("Error decoding base64", err)
		os.Exit(1)
	}

	result = string(decoded)
	return result
}

func getRequest() []byte {
	filepath := "/tmp/opm.json"
	if _, err := os.Stat(filepath); err == nil {
		content, err := os.ReadFile(filepath)
		if err != nil {
			fmt.Println("Error reading file: ", filepath, err)
			os.Exit(1)
		}

		fmt.Printf("INFO: File found at %s, imma use it!\n\n", filepath)
		return content
	}

	response, err := http.Get(opmURL)
	if err != nil {
		fmt.Println("Error: requesting main json file!", err)
		os.Exit(1)
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		fmt.Println("Error: reading response body!")
		os.Exit(1)
	}

	err = os.WriteFile(filepath, body, 0644)
	if err != nil {
		fmt.Println("Error saving response body to file: ", filepath, err)
		os.Exit(1)
	}

	return body
}

func parseResponse() Opm {
	var opm Opm
	err := json.Unmarshal(getRequest(), &opm)
	if err != nil {
		fmt.Println("Error at parsing json response!", err)
		os.Exit(1)
	}

	return opm
}

type Opm struct {
	Chapters    map[string]Chapter `json:"chapters"`
	Cover       string
	Description string
	Title       string
}

type Chapter struct {
	Groups      map[string][]string `json:"groups"`
	LastUpdated int64               `json:"last_updated"`
	Title       string              `json:"title"`
	Volume      string              `json:"volume"`
}
